src/main/java/org/benetech/servicenet/matching/counter/LocationSimilarityCounter.java

```diff
@@ -34,19 +38,55 @@
 
     @Override
     public float countSimilarityRatio(Location location1, Location location2, MatchingContext context) {
-        if (location1 == null || location1.getPhysicalAddress() == null
-            || location2 == null || location2.getPhysicalAddress() == null) {
+        if (anyPhysicalAddressIsNull(location1, location2)) {
             return NO_MATCH_RATIO;
         }
+
+        if (allCoordinatesArePresent(location1, location2)) {
+            return countRatioIfCoordinatesArePresent(location1, location2);
+        } else {
+            return countRatioIfCoordinatesAreMissing(location1, location2, context);
+        }
+    }
+
+    private float countRatioIfCoordinatesAreMissing(Location location1, Location location2, MatchingContext context) {
         float max = NO_MATCH_RATIO;
-        for (GeocodingResult result1 : geocodingResultService.findAllForLocationOrFetchIfEmpty(location1, context)) {
-            for (GeocodingResult result2 : geocodingResultService.findAllForLocationOrFetchIfEmpty(location2, context)) {
+
+        for (GeocodingResult result1 : getGeoCodingResult(location1, context)) {
+            for (GeocodingResult result2 : getGeoCodingResult(location2, context)) {
```

Comment:
How many API calls can be executed in the worst scenario?

Reply:
two, if somehow the result was not saved in the db during the import. However it should be saved, so We're searching the db here. In fact, I should not search it in the nested loop, so I'll extract it.